test/metabase/lib/card_test.cljc

```diff
@@ -91,3 +94,62 @@
              (mapv :name (get-in lib.tu/mock-cards [:orders :result-metadata]))))
       (is (= ["ID" "SUBTOTAL" "TOTAL" "TAX" "DISCOUNT" "QUANTITY" "CREATED_AT" "PRODUCT_ID" "USER_ID"]
              (mapv :name (lib/visible-columns venues-query)))))))
+
+(deftest ^:parallel returned-columns-31769-test
+  (testing "Cards with joins should return correct column metadata/refs (#31769)"
+    (let [metadata-provider (lib.tu.mocks-31769/mock-metadata-provider meta/metadata-provider meta/id 1 2)
+          card              (lib.metadata/card metadata-provider 1)
+          q                 (:dataset-query card)
+          cols              (lib/returned-columns q)]
+      (is (=? [{:name                         "CATEGORY"
+                :lib/source                   :source/breakouts
+                :lib/source-column-alias      "CATEGORY"
+                :metabase.lib.join/join-alias "Products"
+                :lib/desired-column-alias     "Products__CATEGORY"}
+               {:name                     "count"
+                :lib/source               :source/aggregations
+                :lib/source-column-alias  "count"
+                :lib/desired-column-alias "count"}]
+              cols))
+      (is (=? [[:field {:join-alias "Products"} (meta/id :products :category)]
+               [:aggregation {} string?]]
+              (map lib.ref/ref cols))))))
+
+(deftest ^:parallel returned-columns-31769-source-card-test
+  (testing "Queries with `:source-card`s with joins should return correct column metadata/refs (#31769)"
+    (let [metadata-provider (lib.tu.mocks-31769/mock-metadata-provider)
+          card            (lib.metadata/card metadata-provider 1)
+          q                 (lib/query metadata-provider card)
+          cols              (lib/returned-columns q)]
+      (is (=? [{:name                     "CATEGORY"
+                :lib/source               :source/card
+                :lib/source-column-alias  "CATEGORY"
+                :lib/desired-column-alias "Products__CATEGORY"}
+               {:name                     "count"
+                :lib/source               :source/card
+                :lib/source-column-alias  "count"
+                :lib/desired-column-alias "count"}]
+              cols))
+      (is (=? [[:field {:base-type :type/Text} "Products__CATEGORY"]
+               [:field {:base-type :type/Integer} "count"]]
+              (map lib.ref/ref cols))))))
```

Comment:
Here I understand even less (if that's possible) why `:name` is `"CATEGORY"`. And if we do look into the contents of the query, why do we throw away `:metabase.lib.join/join-alias`?

Reply:
`:name` isn't supposed to change at all, it's supposed to be the original name of the column i.e. `metabase_field.name`. So it should always be `CATEGORY` regardless of where it came from.

The join alias is not relevant outside of the stage where the join happens, subsequent stages aren't supposed to need to know where a Field came from in order to use it. Any stage after one where a field is first introduced should always be using nominal field references e.g. `[:field {} "field_name"]`, see https://github.com/metabase/metabase/issues/29763. But completely fixing this is going to have to wait until we port more of the QP and FE over I think because there are a lot of broken metadata/queries out in the wild those both have their own way of dealing with this stuff, we don't want to break something there

Reply:
Why should `:name` stay unchanged? I feel that just leaks information causing confusion later. I mean, once we close the stage, the next stage should see the alias we gave it, shouldn't it? Or is this what you mean cannot be fixed just yet?

Reply:
`:name` stays unchanged because it's unchanged in QP results metadata and changing it will break things (I tried this already, and we need to maintain compatibility with MLv1/QP behavior for now until we port everything).

`:lib/source-column-alias` and `:lib/desired-column-alias` change between stages, so those are what you should be looking at.

Reply:
What does "subsequent" stage mean? Any stage other than the first?

Reply:
yeah that's what I meant

Reply:
Wouldn't it be best always setting `:lib/desired-column-alias`, even if it's the same as `:name`?

Reply:
I agree with that last. I think it's already set much of the time, at least it tends to be returned by `visible-columns` and `returned-columns`.

Also this sounds like we should be doing substantially less fuzzy matching in `find-closest-matches-for-refs` et al, most of the time. The matching feels like it should proceed more like this:
- If the needle ref has a numeric ID, match precisely by that.
    - If that doesn't find anything, and we have a `query`, look up the numeric ID in that context and find its `:name`.
- With a string ref, match precisely by that (or with `(or (:lib/desired-column-alias col) (:name col))` I guess if `:lib/desired-column-alias` isn't always present).

Then, if either stage of that matching produces > 1 match, then we can use eg. `:join-alias` and `:temporal-unit` for disambiguation, to find eg. the correct joined column for cases where two FKs point at the same table.

That would be both more precise and likely much faster.